Tests/GRPCHTTP2CoreTests/GRPCStreamStateMachineTests.swift

```diff
@@ -0,0 +1,2118 @@
+    ])
+    try stateMachine.receive(message: receivedBytes, endStream: false)
+    
+    // Close server
+    XCTAssertNoThrow(try stateMachine.receive(metadata: .init(), endStream: true))
+    
+    let receivedMessage = try XCTUnwrap(stateMachine.nextInboundMessage())
+    XCTAssertEqual(receivedMessage, [42, 42])
+    
+    XCTAssertNil(stateMachine.nextInboundMessage())
+  }
+  
+  func testNextInboundMessageWhenClientClosedAndServerIdle() {
+    var stateMachine = makeClientStateMachine()
+    
+    // Open client
+    XCTAssertNoThrow(try stateMachine.send(metadata: self.testMetadata))
+    
+    // Close client
+    XCTAssertNoThrow(try stateMachine.send(message: [], endStream: true))
+    
+    // If server is idle it means we never got any messages, assert no inbound
+    // message is present.
+    XCTAssertNil(stateMachine.nextInboundMessage())
+  }
+  
+  func testNextInboundMessageWhenClientClosedAndServerOpen() throws {
+    var stateMachine = makeClientStateMachine()
+    
+    // Open client
+    XCTAssertNoThrow(try stateMachine.send(metadata: self.testMetadata))
+    
+    // Open server
+    XCTAssertNoThrow(try stateMachine.receive(metadata: .init(), endStream: false))
+    
+    let receivedBytes = ByteBuffer(bytes: [
+      0, // compression flag: unset
+      0, 0, 0, 2, // message length: 2 bytes
+      42, 42 // original message
+    ])
+    try stateMachine.receive(message: receivedBytes, endStream: false)
+    
+    // Close client
+    XCTAssertNoThrow(try stateMachine.send(message: [], endStream: true))
+    
+    // Even though the client is closed, because it received a message while open,
+    // we must get the message now.
+    let receivedMessage = try XCTUnwrap(stateMachine.nextInboundMessage())
+    XCTAssertEqual(receivedMessage, [42, 42])
+    
+    XCTAssertNil(stateMachine.nextInboundMessage())
+  }
+  
+  func testNextInboundMessageWhenClientClosedAndServerClosed() throws {
+    var stateMachine = makeClientStateMachine()
+    
+    // Open client
+    XCTAssertNoThrow(try stateMachine.send(metadata: self.testMetadata))
+    
+    // Open server
+    XCTAssertNoThrow(try stateMachine.receive(metadata: .init(), endStream: false))
+    
+    let receivedBytes = ByteBuffer(bytes: [
+      0, // compression flag: unset
+      0, 0, 0, 2, // message length: 2 bytes
+      42, 42 // original message
+    ])
+    try stateMachine.receive(message: receivedBytes, endStream: false)
+    
+    // Close server
+    XCTAssertNoThrow(try stateMachine.receive(metadata: .init(), endStream: true))
+    
+    // Close client
+    XCTAssertNoThrow(try stateMachine.send(message: [], endStream: true))
+    
+    // Even though the client is closed, because it received a message while open,
+    // we must get the message now.
+    let receivedMessage = try XCTUnwrap(stateMachine.nextInboundMessage())
+    XCTAssertEqual(receivedMessage, [42, 42])
+    
+    XCTAssertNil(stateMachine.nextInboundMessage())
+  }
+}
```

Comment:
All of these tests cover performing an action in each state (which is great). I think we need a second kind of test which exercises common paths through the state machine, e.g.:

1. client sends metadata, client sends message(s), client sends end, server sends metadata, server sends message(s), server sends end
2. client sends metadata, server sends metadata, ... etc.

This is effectively like picking parts of individual tests above and combining them but I think is still valuable as you can exercise different routes through the state machine and is closer to how the state machine will actually be used.